Features: Login, Show Attributes and Issuance

quincydejong opened the pull request:
This PR adds the following features for users:
* Users can log in at external identity providers to get attributes
* Users can see the attributes they got from identity providers before loading it into their Yivi app
* Users can see load these attributes in the Yivi app
* Users will see a confirmation page if successful

(This closes US02 and US03)

quincydejong committed:
feat: add .env

quincydejong committed:
chore: add Microsoft.AspNetCore.Authentication.OpenIdConnect

quincydejong committed:
chore: remove empty App.css

quincydejong committed:
refactor: App.tsx to LoginPage.tsx

quincydejong committed:
fix: remove template logo's

quincydejong committed:
feat: add SchemeCredentialClient

quincydejong committed:
feat: add IrmaServerClient

quincydejong committed:
chore: remove unnecessary interface for IdentityProviderService

quincydejong committed:
docs: add comments for the FileSystem classes

quincydejong committed:
chore: add Microsoft.AspNetCore.Authentication.OpenIdConnect

quincydejong committed:
feat: add duplicate slug key check

quincydejong committed:
feat: update identity provider configuration

Also updates tests.

quincydejong committed:
feat: add IdentityProvidersExtension

quincydejong committed:
test: add a duplicate slug test

quincydejong committed:
feat: add a method to get all identity providers

quincydejong committed:
feat: add AttributeMapperService

quincydejong committed:
feat: add ApiClient

quincydejong committed:
test: add initial AttributeMapperServiceTests

quincydejong committed:
feat: add CredentialAttributeService

quincydejong committed:
feat: add endpoints in IdentityProviderController

For logging in, human-readable attributes and credential issuance.

quincydejong committed:
chore: add commonjs support for Yivi frontend package

quincydejong committed:
fix: remove incorrect exception

quincydejong committed:
chore: add commonjs plugin for Yivi frontend package to package.json

quincydejong committed:
feat: add UI translations

Includes some small refactoring.

quincydejong committed:
feat: add initial Issuance page

quincydejong committed:
feat: add initial Done page

quincydejong committed:
feat: add routing to attributes and success page

quincydejong committed:
wip: temp fix for Yivi QR code

kamphuisem commented on GIAP.Server/.gitignore:
File can be removed

kamphuisem commented on GIAP.Server/Controllers/IdentityProviderController.cs:
```diff
@@ -1,19 +1,191 @@
 [ApiController]
-public class IdentityProviderController(IIdentityProviderService identityProviderService) : ControllerBase
+public class IdentityProviderController(
+    IdentityProviderService identityProviderService,
```
Usually, in controllers you want to inject interfaces, rather then classes.
The interfaces should be registered with the IoC-container (ie. `services.Add[Scoped|Transient|Singleton]<...>`).
Otherwise, the controller is difficult to add tests for.

kamphuisem commented on GIAP.Server/Controllers/IdentityProviderController.cs:
```diff
@@ -1,19 +1,191 @@
+        if (failureResponse != null || idpAuthData == null) return failureResponse!;
+
+        // Get scheme attributes
+        var schemeUrl = DotNetEnv.Env.GetString("SCHEME_BASE_URL");
```
`DotNetEnv.Env` is a static call, which makes it impossible to mock/stub if you want to add tests.
It would be better to create a wrapper class around the DotNetEnv stuff, and inject that in the controller.

quincydejong committed:
cicd: add workflow_dispatch in delivery.yml

kamphuisem commented on GIAP.Server/Controllers/IdentityProviderController.cs:
```diff
@@ -1,19 +1,191 @@
+    public async Task<IActionResult> IrmaSessionStart(string slug)
+    {
+        var (failureResponse, idpAuthData) = await IdentityProviderAuth(HttpContext, slug);
+        if (failureResponse != null || idpAuthData == null) return failureResponse!;
```
This will fail if both `failureResponse` and `idpAuthData` are null,

kamphuisem commented on GIAP.Server/Controllers/IdentityProviderController.cs:
```diff
@@ -1,19 +1,191 @@
+    public async Task<IActionResult> GetAttributes(string slug, string language)
+    {
+        var (failureResponse, idpAuthData) = await IdentityProviderAuth(HttpContext, slug);
+        if (failureResponse != null || idpAuthData == null) return failureResponse!;
```
This will fail if both failureResponse and idpAuthData are null

kamphuisem commented on GIAP.Server/Controllers/IdentityProviderController.cs:
```diff
@@ -1,19 +1,191 @@
+    /// <param name="httpContext">The HTTP context to log in the user at the identity provider.</param>
+    /// <param name="slug">The slug of the identity provider.</param>
+    /// <returns>Returns a tuple with either a failure response or IdentityProviderAuthData if successful.</returns>
+    private async Task<(IActionResult? failureResponse, IdentityProviderAuthData? authIdpData)> IdentityProviderAuth(
```
I would suggest using `out` parameters for the data.
Now you have kind of a weird construction using anonymous tuples (which is not really C# prettiest way of programming)

kamphuisem commented on GIAP.Server/Services/ApiClient.cs:
```diff
@@ -0,0 +1,59 @@
+
+        var response = await httpClient.GetAsync(url);
+
+        response.EnsureSuccessStatusCode();
```
What will happen if the api didn't return a success-status?

kamphuisem commented on GIAP.Server/Services/CredentialAttributeService.cs:
```diff
@@ -0,0 +1,57 @@
+/// Service for getting credential attributes for issuance and display.
+/// </summary>
+/// <param name="attributeMapperService">The attribute mapper.</param>
+public class CredentialAttributeService(AttributeMapperService attributeMapperService)
```
Inject interfaces, not classes.

kamphuisem commented on GIAP.Server/Services/IdentityProviderService.cs:
```diff
@@ -8,7 +8,7 @@ namespace GIAP.Server.Services;
 /// Class is used as a singleton, see Program.cs.
 /// </summary>
 /// <param name="fileSystem">File system used to read the identity providers file.</param>
-public class IdentityProviderService(IFileSystem fileSystem) : IIdentityProviderService
```
Why did you remove the interface?
It is still injected somewhere, right?

quincydejong committed:
fix: add missing `workflow_dispatch:`

kamphuisem commented on GIAP.Server/Services/IdentityProviderService.cs:
```diff
@@ -56,4 +68,10 @@ public void Initialize()
+    /// Get all identity providers.
+    /// </summary>
+    /// <returns>A list of all the identity providers.</returns>
+    public List<IdentityProvider> GetAll() => _identityProviders;
```
Return an `IReadOnlyList<..>` as you don't want other objects to changes the contents of this list.

kamphuisem commented on GIAP.Server/Services/IrmaServerClient.cs:
```diff
@@ -0,0 +1,61 @@
+        var irmaServerUrl = irmaServerBaseUrl + "/session";
+        var validity = DateTimeOffset.UtcNow.AddMonths(issuanceValidityInMonths).ToUnixTimeSeconds();
+
+        var requestData = new Dictionary<string, object>
```
Why use a Dictionary for this object?
Why not create a typed `IssuanceRequest` class?

kamphuisem commented on GIAP.Server/Services/IrmaServerClient.cs:
```diff
@@ -0,0 +1,61 @@
+        };
+
+        var response = await httpClient.PostAsJsonAsync(irmaServerUrl, requestData);
+        response.EnsureSuccessStatusCode();
```
What if not successful?

kamphuisem commented on GIAP.Server/Services/SchemeCredentialClient.cs:
```diff
@@ -0,0 +1,45 @@
+/// The SchemeCredentialClient is used to get credential from the scheme.
+/// </summary>
+/// <param name="httpClient">The HTTP client to call the scheme endpoint.</param>
+public class SchemeCredentialClient(HttpClient httpClient)
```
Usually, IRMA-servers keep a cache of the scheme locally, and handle everything using that cache.
Did you create an issue on the backlog to change this functionality?

kamphuisem commented on giap.client/src/pages/Issuance.tsx:
```diff
@@ -0,0 +1,105 @@
+            // Back-end options
+            session: {
+                // Point this to your controller:
+                url: `http://localhost:62858/api/identity-provider/${slug}/irma-endpoint`,
```
How does this work on other environments then localhost?

quincydejong committed:
Merge branch 'main' into feature/us02-us03-login-attributes-issuance

quincydejong committed:
wip: fix http callback issue

quincydejong committed:
wip: fix http callback issue attempt 2

quincydejong committed:
wip: fix http callback issue part 3

quincydejong committed:
wip: fix http callback part 4

quincydejong committed:
wip: fix http callback part 5

quincydejong committed:
wip: fix http callback part 6

quincydejong committed:
wip: fix http callback part 7

quincydejong committed:
wip: fix 502 attempt 1

quincydejong committed:
wip: fix 502 attempt 2

quincydejong committed:
wip: fix 502 attempt 3

quincydejong committed:
wip: fix 502 attempt 4

quincydejong committed:
wip: fix 502 attempt 5

quincydejong committed:
chore: remove empty .gitignore

quincydejong committed:
feat: add IAttributeMapperService for testing

quincydejong committed:
feat: add ICredentialAttributeService for testing

quincydejong committed:
fix: add missing registration and usage of added interfaces

quincydejong committed:
feat: add IApiClient for testing

quincydejong committed:
docs: add missing docs for interfaces

quincydejong committed:
feat: add ISchemeCredentialClient for testing

quincydejong committed:
feat: add IIrmaServerClient for testing

quincydejong committed:
revert: "chore: remove unnecessary interface for IdentityProviderService"

This reverts commit db8ab49f8b7fd76159635bd169c773281b8f6b86.

quincydejong committed:
feat: add IIdentityProviderService for testing

quincydejong committed:
feat: change List to IReadOnlyList for IdPService

quincydejong committed:
wip: add additional ForwardedHeaders

quincydejong committed:
fix: get IdPService through interface

quincydejong committed:
wip: add redirect logging

quincydejong committed:
wip: temp disable SaveTokens

quincydejong committed:
wip: turning SaveTokens back on

quincydejong committed:
wip: use server memory for cookies

quincydejong committed:
wip: debug invalid request URI IrmaServer part 1

quincydejong committed:
Revert "wip: use server memory for cookies"

This reverts commit d6772f267eaf98c981b02195bf5e8841236df9d4.

quincydejong committed:
wip: remove cookie chunking

quincydejong committed:
fix: add cookie back

quincydejong committed:
refactor: remove temp HeaderSize debug logging

quincydejong committed:
feat: add dynamic environment url

`window.location.origin` can point to http:localhost:62858 but can also point to `https://giap.staging.yivi.app/`

quincydejong committed:
refactor: replace auth method to auth middleware

quincydejong committed:
feat: add DotNetEnv wrapper

quincydejong committed:
feat: add typed IssuanceResponse class

quincydejong committed:
feat: add typed IssuanceRequest

quincydejong committed:
refactor: remove the need for JsonSerializerOptions in IrmaServerClient

This also ensures we get the same JsonPropertyNames for the frontend. (Although this didn't seem to be an issue.)

quincydejong committed:
refactor: remove todo and small refactor

quincydejong committed:
wip: investigate the need for Redirect instead of LocalRedirect

quincydejong committed:
docs: remove temp todo's

quincydejong committed:
fix: health check does not require authentication

quincydejong committed:
fix: middleware order health endpoint check

quincydejong committed:
fix: allow non-backend requests to pass through without auth

quincydejong committed:
docs: moved comment to the backlog

quincydejong committed:
docs: removed temp todo and temp debugging

quincydejong committed:
docs: update UiLogin docs

quincydejong committed:
fix: out of sync cookie and access token lifetime

quincydejong committed:
fix: remove the white border around the image

quincydejong committed:
refactor: align the content to be centered

Based on feedback, the contents of the page are now centered.
Also removes the dependency of absolute CSS positioning.

quincydejong committed:
feat: update translations

quincydejong committed:
refactor: remove temp logs

quincydejong committed:
refactor: remove .'s in the header of the Done.tsx page

quincydejong committed:
feat: add server error handling and client logging

quincydejong committed:
refactor: change loginButton text based on feedback

Feedback from team demo 16/06/2025

quincydejong committed:
revert: styling changes back to 'left align' based on feedback

Feedback from team demo 16/06/2025: styling should match other Yivi demos.

quincydejong committed:
fix: remove unnecessary CSS

quincydejong committed:
feat: add dynamic language support for the Yivi client component

quincydejong committed:
style: add slight padding to the table to make it more readable

kamphuisem commented on GIAP.Server/Configuration/IdentityProvidersExtension.cs:
```diff
@@ -0,0 +1,40 @@
+                configure.ClientSecret = identityProvider.ClientSecret;
+                configure.CallbackPath = identityProvider.CallbackPath;
+                configure.ResponseType = OpenIdConnectResponseType.Code;
+                configure.SaveTokens = true; // necessary to get the access token for optional API calls
```
Thought: shouldn't this be a setting in the config file?
If there is no additional API calls to be made, it doesn't really make sense to save the tokens, right?
Maybe create an issue if you agree.

kamphuisem commented on GIAP.Server/Controllers/IdentityProviderController.cs:
```diff
@@ -4,16 +4,173 @@
+    {
+        try
+        {
+            var idpAuthData = (HttpContext.Items["authIdpData"] as IdentityProviderAuthData)!; // trust middleware
```
I would avoid using the `!` syntax and opt for a normal `if == null` check and return a `Unauthorized` message.
If there is a bug in the middleware, or the middleware isn't configured (correctly), this will cause unexpected behaviour in the next lines of the function.
The `if` check allows for more control over the situation.

kamphuisem commented on GIAP.Server/Controllers/IdentityProviderController.cs:
```diff
@@ -4,16 +4,173 @@
+        }
+        catch (HttpRequestException)
+        {
+            return StatusCode(502, "Bad Gateway: Unable to get attributes from external services.");
```
Your application doesn't really act as a gateway, so this code doesn't make sense.
I would opt to change it to a generic `500` message.

kamphuisem commented on GIAP.Server/Controllers/IdentityProviderController.cs:
```diff
@@ -4,16 +4,173 @@
+        }
+        catch (HttpRequestException)
+        {
+            return StatusCode(502,
```
See previous comment

kamphuisem commented on GIAP.Server/Controllers/IdentityProviderController.cs:
```diff
@@ -4,16 +4,173 @@
+    {
+        try
+        {
+            var idpAuthData = (HttpContext.Items["authIdpData"] as IdentityProviderAuthData)!; // trust middleware
```
See previous comment

kamphuisem commented on GIAP.Server/Middleware/IdentityProviderAuthMiddleware.cs:
```diff
@@ -0,0 +1,91 @@
+
+    private string GetAccessToken(AuthenticateResult userResult)
+    {
+        return userResult.Properties!.GetTokenValue(OpenIdConnectParameterNames.AccessToken)!;
```
Are you 100% sure both `Properties` and `GetTokenValue` cannot ever be null here?

kamphuisem commented on GIAP.Server/Models/IssuanceResponse.cs:
```diff
@@ -0,0 +1,37 @@
+﻿using System.Text.Json.Serialization;
```
Maybe good to add `<seealso ... />` references to the Yivi documentation in your comments

kamphuisem commented on GIAP.Server/Program.cs:
```diff
@@ -24,13 +70,8 @@
-app.UseAuthorization();
-
 app.MapControllers();
 app.MapHealthChecks("/health");
```
Usually we use `UseHealthChecks` instead.
This is a middleware, which can run before the authentication middleware, allowing you to remove the check on `/health` in your custom middleware.
Make sure to call this function _before_ the authorization stuff etc, because the order of adding middleware is important.

kamphuisem commented on GIAP.Server/Services/IdentityProviderService.cs:
```diff
@@ -50,10 +54,9 @@ public void Initialize()
-    /// <param name="slug">A web url slug, for example, "idp-slug".</param>
-    /// <returns>The identity provider, if none found, it returns null.</returns>
+    /// <inheritdoc/>
     public IdentityProvider? GetBySlug(string slug) => _identityProviders.FirstOrDefault(idp => idp.Slug == slug);
```
Maybe always check against lowercase slugs?

kamphuisem commented on GIAP.Server/Services/IdentityProviderService.cs:
```diff
@@ -41,7 +32,20 @@ public void Initialize()
+            foreach (var identityProvider in _identityProviders)
+            {
+                // If a slug is already added to the dictionary, it means a duplicate exists in the idp config file
+                if (slugs.ContainsKey(identityProvider.Slug))
```
What will happen if both `caesar` and `Caesar` and/or `CAESAR` would be configured?
I think that would mean an error right?
I would suggest to keep/match the slugs lowercased.

kamphuisem reviewed (dismissed):
Approved, but please make issues from the comments left

quincydejong committed:
docs: add additional configuration explanation

quincydejong committed:
docs: update frontend and backend development instructions
